cjdns-hdr/src/route_header.rs

```diff
@@ -42,36 +42,33 @@ impl RouteHeader {
     /// * if "[is_ctrl](struct.RouteHeader.html#structfield.is_ctrl) - [public_key](struct.RouteHeader.html#structfield.public_key) - [ip6](struct.RouteHeader.html#structfield.ip6)" invariant is not met;
     /// * if flag for message type states not control, nor incoming frame.
     pub fn parse(data: &[u8]) -> Result<Self, ParseError> {
-        if data.len() != Self::SIZE {
-            return Err(ParseError::InvalidPacketSize);
-        }
         let mut data_reader = Reader::new(data);
-        let public_key = {
-            let public_key_array = data_reader.read_array_32().expect("invalid header data size");
-            if ZERO_PUBLIC_KEY_BYTES == public_key_array {
-                None
-            } else {
-                Some(CJDNSPublicKey::from(public_key_array))
-            }
-        };
-        let switch_header = {
-            let header_bytes = data_reader.take_bytes(SwitchHeader::SIZE).expect("invalid header data size");
-            SwitchHeader::parse(header_bytes)?
-        };
-        let version = data_reader.read_u32_be().expect("invalid header data size");
-        let (is_ctrl, is_incoming) = {
-            let flags = data_reader.read_u8().expect("invalid header data size");
-            (flags & CONTROL_FRAME != 0, flags & INCOMING_FRAME != 0)
+        let (pk_bytes, header_bytes, version, flags, _zeroes, ip6_bytes) = data_reader
+            .read(SizePredicate::Exact(Self::SIZE), |r| {
+                let pk_bytes = r.read_array_32()?;
+                let header_bytes = r.read_slice(SwitchHeader::SIZE)?;
+                let version = r.read_u32_be()?;
+                let flags = r.read_u8()?;
+                // padding
+                let zeroes = r.read_slice(3)?;
```

Comment:
Same, we can just call `read_zeroes(3)` here and do not return the slice at all.

Reply:
discussed. using `skip`